software/oada_upload/src/oada_upload.ts

```diff
@@ -0,0 +1,387 @@
+  // database was loaded with > 300,000 data points. That is 
+  // about a season of gps data but should still be careful.
+  // Going to leave it commented out for now until we determine 
+  // a more efficient way to do this 
+  /*console.log(`Copying time id's not already in the sent db`);
+  await db.query(`SELECT time FROM gps WHERE gps.time NOT IN (SELECT g_time from gps_sent)`);*/
+
+  console.log(`Connecting to OADA: ${domain}`);
+  const oada = await connect({ domain, token });
+  
+  // Infinite loop to continuously check the db for unsent data
+  for (;;) {
+    // Extract x number of unsent data points
+    const gps = await db.query(
+      `SELECT gps.time as time_tz, extract(epoch from gps.time) as time_epoch, gps.lat, gps.lng, gps_sent.sent
+        FROM gps
+        JOIN gps_sent ON gps.time = gps_sent.g_time
+
+        WHERE sent = FALSE
+        ORDER BY time DESC
+        LIMIT $1;`,
+      [batchsize]
+    );
+
+    // If we get no data from the databse, we sent everything.
+    // Wait 1s for the data to start to refill the db and requery
+    if (!gps.rows.length) {
+      console.log('No unsent data found in database');
+      await sleep(1000);
+      continue;
+    }
+    
+    console.info(`Found batch of ${gps.rows.length} GPS points`);
+    console.debug(`Head time: ${gps.rows[0].time_tz}`);
+
+    // Create DataIndex object(?) for filling data in
+    const data: DataIndex = {};
+    // For each datapoint used, extract data and begin to sort them into buckets to be uploaded
+    // Due to how the OADA server is setup, we the buckets will be by hour collected
+    gps.rows.forEach((p) => {
+      //console.debug(`Processing GPS time: ${p.time_tz}`);
+
+      // Extract time and use it to create path that it will be uploaded it
+      const t = moment.unix(p.time_epoch);
+      const day = t.format('YYYY-MM-DD');
+      const hour = t.format('HH');
+      const path = `/bookmarks/isoblue/device-index/${id}/location/day-index/${day}/hour-index/${hour}`;
+      // Create a UUID that can be coarsely sorted by time of creation
+      // TODO: find a way to use the timestamp to create the id instead of the current time
+      const pId = ksuid.randomSync().string;
+
+      // If the current bucket does not exist, create it
+      if (!data[path]) {
+        data[path] = {
+          epochs: [],
+          timetzs: [],
+          locations: {},
+        };
+      }
+
+      // Insert our data into the bucket
+      data[path].epochs.push(p.time_epoch);
+      data[path].timetzs.push(p.time_tz);
+      data[path].locations[pId] = {
+        id: pId,
+        time: {
+          value: p.time_epoch,
+        },
+        location: {
+          lat: p.lat,
+          lng: p.lng,
+        },
+      };
+      
+    });
+    // pEachSeries executes the following code for each bucket (path) that we created sequentially 
+    await pEachSeries(Object.keys(data), async (path) => {
```

Comment:
You can likely drop `pEachSeries` and replace with a `for (let path in data) { ...` because `oada-client` will limit request concurrency. Not sure its a required change though.

Reply:
Err... I guess you'll want something like `await Promise.all(Object.keys(data), (path) => { ... }` instead of the `for (let path in data) { ... ` so that you wait until all of the pending requests to complete before going back and fetching more data from the database.

Reply:
In order to make use of the `oada.client` concurrency we need to do this. Going to try `Promise.all` and `pMap`

Reply:
Promise.all and pMap should be equal if you don't care about the concurrency (which you shouldn't because @oada/client will manage it)